src/rpc/request.cpp

```diff
@@ -131,7 +134,11 @@ bool GetAuthCookie(std::string *cookie_out)
 void DeleteAuthCookie()
 {
     try {
-        fs::remove(GetAuthCookieFile());
+        std::string existing_cookie;
+        if (GetAuthCookie(&existing_cookie) && g_generated_cookie == existing_cookie) {
```

Comment:
If `g_generated_cookie` is not `nullopt`, will `g_generated_cookie == existing_cookie` ever be `false`? I don't think so, and I think that makes this check a bit misleading. I would thus either convert `g_generated_cookie` into a `bool` (as e.g. motivated [here](https://github.com/bitcoin/bitcoin/pull/28946#discussion_r1406275659)), or as suggested by myself in https://github.com/bitcoin/bitcoin/pull/28946#discussion_r1406262955 use a `std::optional<fs::path>` but then perform this check only on the truthiness of `g_generated_cookie`, and not its value.

<details>
<summary>git diff on d95dde9441</summary>

```diff
diff --git a/src/rpc/request.cpp b/src/rpc/request.cpp
index c706ddf8d5..937a3a7dd3 100644
--- a/src/rpc/request.cpp
+++ b/src/rpc/request.cpp
@@ -80,7 +80,7 @@ static fs::path GetAuthCookieFile(bool temp=false)
     return AbsPathForConfigVal(gArgs, arg);
 }
 
-static std::optional<std::string> g_generated_cookie;
+static std::optional<fs::path> g_generated_cookie;
 
 bool GenerateAuthCookie(std::string *cookie_out)
 {
@@ -107,7 +107,7 @@ bool GenerateAuthCookie(std::string *cookie_out)
         LogPrintf("Unable to rename cookie authentication file %s to %s\n", fs::PathToString(filepath_tmp), fs::PathToString(filepath));
         return false;
     }
-    g_generated_cookie = cookie;
+    g_generated_cookie = filepath;
     LogPrintf("Generated RPC authentication cookie %s\n", fs::PathToString(filepath));
 
     if (cookie_out)
@@ -133,12 +133,10 @@ bool GetAuthCookie(std::string *cookie_out)
 
 void DeleteAuthCookie()
 {
+    // Delete the cookie file if it exists and was generated by this process
+    if (!g_generated_cookie) return;
     try {
-        std::string existing_cookie;
-        if (GetAuthCookie(&existing_cookie) && g_generated_cookie == existing_cookie) {
-            // Delete the cookie file if it exists and was generated by this process
-            fs::remove(GetAuthCookieFile());
-        }
+            fs::remove(g_generated_cookie.value());
     } catch (const fs::filesystem_error& e) {
         LogPrintf("%s: Unable to remove random auth cookie file: %s\n", __func__, fsbridge::get_filesystem_error_message(e));
     }

```
</details>

Reply:
Or combining this with a `fs::remove` simplification:

<details>
<summary>git diff on d95dde9441</summary>

```diff
diff --git a/src/rpc/request.cpp b/src/rpc/request.cpp
index c706ddf8d5..06962d2b87 100644
--- a/src/rpc/request.cpp
+++ b/src/rpc/request.cpp
@@ -80,7 +80,7 @@ static fs::path GetAuthCookieFile(bool temp=false)
     return AbsPathForConfigVal(gArgs, arg);
 }
 
-static std::optional<std::string> g_generated_cookie;
+static std::optional<fs::path> g_generated_cookie;
 
 bool GenerateAuthCookie(std::string *cookie_out)
 {
@@ -107,7 +107,7 @@ bool GenerateAuthCookie(std::string *cookie_out)
         LogPrintf("Unable to rename cookie authentication file %s to %s\n", fs::PathToString(filepath_tmp), fs::PathToString(filepath));
         return false;
     }
-    g_generated_cookie = cookie;
+    g_generated_cookie = filepath;
     LogPrintf("Generated RPC authentication cookie %s\n", fs::PathToString(filepath));
 
     if (cookie_out)
@@ -133,14 +133,10 @@ bool GetAuthCookie(std::string *cookie_out)
 
 void DeleteAuthCookie()
 {
-    try {
-        std::string existing_cookie;
-        if (GetAuthCookie(&existing_cookie) && g_generated_cookie == existing_cookie) {
-            // Delete the cookie file if it exists and was generated by this process
-            fs::remove(GetAuthCookieFile());
-        }
-    } catch (const fs::filesystem_error& e) {
-        LogPrintf("%s: Unable to remove random auth cookie file: %s\n", __func__, fsbridge::get_filesystem_error_message(e));
+    // Delete the cookie file if it exists and was generated by this process
+    if (!g_generated_cookie) return;
+    if (std::error_code error; !fs::remove(g_generated_cookie.value(), error) && error) {
+        LogPrintf("%s: Unable to remove random auth cookie file: %s\n", error.message());
     }
 }
 

```
</details>

Reply:
Initially I used a boolean flag (https://github.com/bitcoin/bitcoin/commit/7cb9367157eb42ee06bc6fa024522cc14a80138d) but following [this comment](https://github.com/bitcoin/bitcoin/pull/28784#pullrequestreview-1720677514) it was changed to contain the cookie contents.

I agree that `g_generated_cookie == existing_cookie` can be `false` only if the cookie was somehow overwritten by some other process (i.e. not by `bitcoind`), so indeed it would be simpler to use the patches suggested above.

@luke-jr what do you think?

Reply:
I think it would be better to stick to using a boolean flag since checking the contents currently is not meaningfully useful or different.

Reply:
Sounds good, reverted to use a boolean flag.